sdk/log/batch.go

```diff
@@ -210,12 +210,45 @@ func (b *BatchingProcessor) Shutdown(ctx context.Context) error {
 	return errors.Join(err, b.exporter.Shutdown(ctx))
 }
 
+var errPartialFlush = errors.New("partial flush: export buffer full")
+
+// Used for testing.
+var ctxErr = func(ctx context.Context) error {
+	return ctx.Err()
+}
+
 // ForceFlush flushes queued log records and flushes the decorated exporter.
 func (b *BatchingProcessor) ForceFlush(ctx context.Context) error {
 	if b.stopped.Load() {
 		return nil
 	}
-	err := b.exporter.Export(ctx, b.q.Flush())
+
+	errCh := make(chan error, 1)
+	go func() {
+		defer close(errCh)
+
+		buf := make([]Record, b.q.cap)
+		notFlushed := func() bool {
+			var flushed bool
+			_ = b.q.TryDequeue(buf, func(r []Record) bool {
+				flushed = b.exporter.EnqueueExport(r)
+				return flushed
+			})
+			return !flushed
+		}
+		// For as long as ctx allows, try to make a single flush of the queue.
+		for notFlushed() {
```

Comment:
This can "kill the CPU" (consume a whole CPU core). 

`b.exporter.EnqueueExport(r)` returns `false` and it will continuously loop until `b.exporter.input` is ready to accept data to the channel.

Morevoer, `b.exporter.EnqueueExport(r)` will always return `false` when `Shutdown` is called concurrently. In this scenario `for notFlushed()` will run until the context is done. _For fixing this I think, we could simply return `true` in `b.exporter.EnqueueExport` when the exporter is stopped._

I think we can implement something like:

```go
func (e *bufferExporter) EnqueueExportContext(ctx context.Context, records []Record) bool {
	if len(records) == 0 {
		// Nothing to enqueue, do not waste input space.
		return nil
	}

	data := exportData{ctx: context.Background(), records: records}

	e.inputMu.Lock()
	defer e.inputMu.Unlock()

	// Check stopped before enqueueing now that e.inputMu is held. This
	// prevents sends on a closed chan when Shutdown is called concurrently.
	if e.stopped.Load() {
		return nil // for ForceFlush, returning nil here is good
	}

	select {
	case err := <- ctx.Done():
		return err
	case e.input <- data:
		return nil
	}
}
```

My other idea is to just try once.

Reply:
> This will, again, block TryDequeue until the export is made or the context times out

You are right 🤦 I will look at it later.

> This will consume a goroutine, not necessarily the whole CPU.

This can consume a whole core as there are no sync operations other than `select`. I do not if there is a guarantee that `select` does https://pkg.go.dev/runtime#Gosched when `default` is used. 

> Isn't this acceptable if a user is calling "ForceFlush"

I think that CPU intensive operations should be avoided if possible.

Reply:
> If you are running code on a FaaS system and you have set a timeout on your context but the ForceFlush only attempts a flush once I expect you will be frustrated.

Wouldn't people running in FaaS run Shutdown?

Reply:
I checked (by playing with `GOMAXPROCS=1`) and thanks to `select` the Go runtime can switch to other goroutines.

Moreover, `golangci-lint` does not report `SA5002: this loop will spin, using 100% CPU (staticcheck)`.

Reply:
From my perspective, this conversation can be resolved.

I created a separate comment for concurrent shutdown scenario: https://github.com/open-telemetry/opentelemetry-go/pull/5093#discussion_r1558009790